### PR TITLE
yurt-app-manager support YurtAppDaemon CRD

### PR DESCRIPTION
**Pull request by kadisi**
/kind feature

#### What this PR does / why we need it:

In the edge scenario, compute nodes in different regions are assigned to the same node pool. We expect some workload to be created automatically with the creation of the node pool. No additional human intervention is required.

ref:  https://github.com/openyurtio/openyurt/issues/433
ref:  https://github.com/openyurtio/openyurt/pull/422

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

/assign @rambohe-ch @Fei-Guo 

Please focus on the following files：
/pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go
/yurtappmanager/webhook/yurtappdaemon/mutating/yurtappdaemon_create_update_handler.go
r/pkg/yurtappmanager/webhook/yurtappdaemon/validating/yurtappdaemon_create_update_handler.go

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/yurtappmanager/apis/apps/v1alpha1/uniteddaemonset_types.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
+/*
```
Comment (rambohe-ch) [inline pkg/yurtappmanager/apis/apps/v1alpha1/uniteddaemonset_types.go:1]: file name should revised, uniteddaemonset --> yurtappdaemon

File: pkg/yurtappmanager/webhook/add_yurtappdaemon.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package webhook
```
Comment (rambohe-ch) [inline pkg/yurtappmanager/webhook/add_yurtappdaemon.go:1]: please add openyurt license at the beginning.

File: pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+
+func (r *ReconcileYurtAppDaemon) getOwnedServices(instance *unitv1alpha1.YurtAppDaemon) (*corev1.ServiceList, error) {
+	svcList := &corev1.ServiceList{}
+
+	labelSelector := &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			unitv1alpha1.LabelCurrentYurtAppDaemon: instance.GetName(),
+		},
+	}
+	// 获得YurtAppDaemon 对应的 所有的service
```
Comment (rambohe-ch) [inline pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go:479]: please use english to comment

File: pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+	return len(needCreate) > 0 || len(needDeleted) > 0, utilerrors.NewAggregate(errs)
+}
+
+func (r *ReconcileYurtAppDaemon) classifyWorkloads(instance *unitv1alpha1.YurtAppDaemon, currentNodepoolToWorkload map[string]*workloadcontroller.Workload,
+	allNameToNodePools map[string]unitv1alpha1.NodePool, expectedRevision string) (needDeleted, needUpdate []*workloadcontroller.Workload,
+	needCreate []string) {
+
+	for npName, load := range currentNodepoolToWorkload {
+		find := false
+		for vnp, np := range allNameToNodePools {
```
Comment (rambohe-ch) [inline pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go:354]: how about `if np, ok = allNameToNodePools[vnp]; ok {`

File: pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+		if !find {
+			needDeleted = append(needDeleted, load)
+			klog.V(4).Infof("YurtAppDaemon[%s/%s] need delete [%s/%s/%s]", instance.GetNamespace(),
+				instance.GetName(), load.GetKind(), load.Namespace, load.Name)
+		}
+	}
+
+	for vnp, _ := range allNameToNodePools {
+		find := false
+		for np, _ := range currentNodepoolToWorkload {
```
Comment (rambohe-ch) [inline pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go:388]: the same as above comment, use `currentNodepoolToWorkload[vnp]`

File: pkg/yurtappmanager/webhook/yurtappdaemon/mutating/yurtappdaemon_create_update_handler.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	unitv1alpha1.SetDefaultsYurtAppDaemon(obj)
+	obj.Status = unitv1alpha1.YurtAppDaemonStatus{}
+
+	statefulSetTemp := obj.Spec.WorkloadTemplate.StatefulSetTemplate
+	deployTem := obj.Spec.WorkloadTemplate.DeploymentTemplate
+
```
Comment (rambohe-ch) [inline pkg/yurtappmanager/webhook/yurtappdaemon/mutating/yurtappdaemon_create_update_handler.go:50]: how about add verification that statefulset template and deployment template should not be exist at the same time.

File: pkg/yurtappmanager/apis/apps/v1alpha1/uniteddaemonset_types.go (hunk 1 above)
Comment (kadisi) [reply pkg/yurtappmanager/apis/apps/v1alpha1/uniteddaemonset_types.go]: done

File: pkg/yurtappmanager/webhook/yurtappdaemon/validating/yurtappdaemon_create_update_handler.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+func (h *YurtAppDaemonCreateUpdateHandler) SetOptions(options webhookutil.Options) {
+	h.Client = options.Client
+	return
+}
+
+// Handle handles admission requests.
+func (h *YurtAppDaemonCreateUpdateHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	obj := &unitv1alpha1.YurtAppDaemon{}
+
+	klog.V(2).Infof("prepare to valid united daemonset ...")
```
[the remark this answers is not in the document]
Comment (kadisi) [reply pkg/yurtappmanager/webhook/yurtappdaemon/validating/yurtappdaemon_create_update_handler.go]: done

File: pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go (hunk 3 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go]: done

File: pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go (hunk 4 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go]: done

File: pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go (hunk 5 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/yurtappdaemon/yurtappdaemon_controller.go]: done

File: pkg/yurtappmanager/webhook/add_yurtappdaemon.go (hunk 2 above)
Comment (kadisi) [reply pkg/yurtappmanager/webhook/add_yurtappdaemon.go:17]: done

**Comment by kadisi** [comment]
@rambohe-ch PTAL

**Review by rambohe-ch — APPROVED** [review]

**Pull request merged**